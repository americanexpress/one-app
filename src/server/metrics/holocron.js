/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createCounter } from './counters';
import { createGauge } from './gauges';

import createMetricNamespace from './create-metric-namespace';

const holocronNamespace = createMetricNamespace('holocron');

createCounter({
  name: holocronNamespace('module_map_poll', 'total'),
  help: 'total times the module map has been polled',
});

createCounter({
  name: holocronNamespace('module_map_updated', 'total'),
  help: 'total times the module map has been updated',
});

createCounter({
  name: holocronNamespace('module_map_poll_restarted', 'total'),
  help: 'total times the module map poll has been restarted',
});

createGauge({
  name: holocronNamespace('module_map_poll_wait', 'seconds'),
  help: 'delay until the next poll of the module map',
});

createGauge({
  name: holocronNamespace('module_map_poll_consecutive_errors', 'total'),
  help: 'how many times has the server encountered errors polling the module map',
});

export default holocronNamespace.getMetricNames();
