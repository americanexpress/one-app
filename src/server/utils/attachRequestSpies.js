/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import https from 'node:https';
import http from 'node:http';
import url from 'node:url';
import onFinished from 'on-finished';

export function attachSpy(obj, methodName, spy) {
  const origMethod = obj[methodName];

  if (typeof origMethod !== 'function') {
    throw new TypeError(`${methodName} is not a function`);
  }

  if (typeof spy !== 'function') {
    throw new TypeError(`spy must be a function (was "${typeof spy}")`);
  }

  // we're monkeypatching, we need to reassign a property of the obj argument
  // eslint-disable-next-line no-param-reassign -- see above comment
  obj[methodName] = function spiedMethod(...args) {
    let returnValue;
    let originalCalled = false;
    const callOriginal = () => {
      originalCalled = true;
      returnValue = origMethod.apply(obj, args);
      return returnValue;
    };

    spy([...args], callOriginal);

    if (!originalCalled) {
      callOriginal();
    }

    return returnValue;
  };
}

function buildUrlObject(options, defaultProtocol) {
  // TODO: url.parse is deprecated, use new URL() instead
  // this is not a 1:1 replacement and will require changes.
  // Currently the parsed url is used to provide href to
  // loggers outgoingRequestEndSpy function.
  // https://github.com/americanexpress/one-app/blob/main/src/server/utils/logging/setup.js#L42
  // https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost
  const parsedPath = url.parse(options.path || '/');
  const protocol = options.protocol || `${defaultProtocol}:`;
  const urlObject = {
    auth: options.auth,
    hostname: options.hostname || options.host || 'localhost',
    protocol,
    port: options.port || (protocol === 'http:' ? 80 : 443),
    hash: parsedPath.hash,
    path: parsedPath.path,
    pathname: parsedPath.pathname,
    query: parsedPath.query,
    search: parsedPath.search,
  };
  if (
    (protocol === 'http:' && urlObject.port === 80)
    || (protocol === 'https:' && urlObject.port === 443)
  ) {
    urlObject.port = undefined;
  }
  return urlObject;
}

function parseUrl(options, defaultProtocol) {
  // some data is not stored in the clientRequest, have to duplicate some parsing
  // adapted from https://github.com/nodejs/node/blob/894203dae39c7f1f36fc6ba126bb5d782d79b744/lib/_http_client.js#L22
  if (typeof options === 'string') {
    return url.parse(options);
  }

  return url.parse(url.format(buildUrlObject(options, defaultProtocol)));
}

function httpSpy(defaultProtocol, requestSpy, socketCloseSpy) {
  return (args, callOriginal) => {
    const options = args[0];
    const clientRequest = callOriginal();
    const parsedUrl = parseUrl(options, defaultProtocol);

    requestSpy(clientRequest, parsedUrl);

    if (socketCloseSpy) {
      onFinished(clientRequest, () => socketCloseSpy(clientRequest, parsedUrl));
    }
  };
}

export default function attachRequestSpies(requestSpy, socketCloseSpy) {
  if (typeof requestSpy !== 'function') {
    throw new TypeError(`requestSpy must be a function (was "${typeof requestSpy}")`);
  }

  if (socketCloseSpy && (typeof socketCloseSpy !== 'function')) {
    throw new TypeError(
      `socketCloseSpy must be function if provided (was "${typeof socketCloseSpy}")`);
  }

  attachSpy(https, 'request', httpSpy('https', requestSpy, socketCloseSpy));
  attachSpy(http, 'request', httpSpy('http', requestSpy, socketCloseSpy));
}
