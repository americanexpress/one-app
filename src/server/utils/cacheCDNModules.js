/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import fs from 'fs';
import chalk from 'chalk';

export const getUserHomeDirectory = () => process.env.HOME || process.env.USERPROFILE;
export const cacheFileName = '.one-app-module-cache';
export const moduleCacheFileName = '.one-app-module-map-cache';
export const oneAppDirectoryName = '.one-app';
export const oneAppDirectoryPath = path.join(getUserHomeDirectory(), oneAppDirectoryName);
export const oneAppModuleCachePath = path.join(oneAppDirectoryPath, cacheFileName);

// show cache size and how to delete info on start
export const showCacheInfo = () => {
  fs.stat(oneAppModuleCachePath, (error, stats) => {
    if (error) {
      console.error('There was error checking file stat', error);
    } else {
      const fileSizeOnMB = stats.size / (1024 * 1024); // bytes to mb
      const message = `File size of ${cacheFileName}: ${chalk.bold.greenBright(fileSizeOnMB.toFixed(2), 'MB')}`;
      const separator = '*'.repeat(message.length);
      console.log(chalk.bold.cyanBright(separator));
      console.log(chalk.bold.redBright('CACHE INFORMATION'));
      console.log(message);
      console.log(`To delete cache, please run \n  ${chalk.bold.redBright('  rm ', oneAppModuleCachePath)}`);
      console.log(chalk.bold.cyanBright(separator));
    }
  });
};

// setup folder and file
export const setupCacheFile = () => {
  fs.mkdir(oneAppDirectoryPath, { recursive: true }, (error) => {
    if (error) {
      console.error(`There was error creating ${oneAppDirectoryName} directory`);
    } else {
      console.log(`Successfully created ${oneAppDirectoryPath}`);
      console.log(`Creating ${cacheFileName}`);
      try {
        fs.writeFileSync(oneAppModuleCachePath, JSON.stringify('{}'));
        console.log(`${cacheFileName} created successfully on ${oneAppModuleCachePath}`);
      } catch (err) {
        console.error(`Error creating ${cacheFileName} on ${oneAppModuleCachePath}, \n${err}`);
      }
    }
  });
};

// gets cached module from ~/.one-app/.one-app-module-cache
export const getCachedModules = () => {
  let hasCachedFile = false;
  if (fs.existsSync(oneAppModuleCachePath)) {
    hasCachedFile = true;
  } else {
    setupCacheFile();
  }
  if (hasCachedFile) {
    try {
      showCacheInfo();
      const cachedContent = fs.readFileSync(oneAppModuleCachePath, 'utf8');
      return JSON.parse(cachedContent);
    } catch (err) {
      console.error('Could not parse JSON content', err);
    }
  }
  return {};
};

let timerId = null;

export const writeToCache = (content, delay = 500) => {
  // added debounce
  clearTimeout(timerId);
  timerId = setTimeout(() => {
    fs.writeFile(oneAppModuleCachePath, JSON.stringify(content, null, 2), (error) => {
      if (error) {
        console.log(`There was an error updating content \n ${error}`);
      }
    });
    timerId = null;
  }, delay);
};

export const removeDuplicatedModules = (url, cachedModules, moduleNames) => {
  const matchingModule = moduleNames.find((moduleName) => url.match(new RegExp(`\\b\\/${moduleName}\\/\\b`)));

  const updatedCachedModules = cachedModules;
  Object.keys(updatedCachedModules).forEach((cachedModuleKey) => {
    if (cachedModuleKey.match(new RegExp(`\\b\\/${matchingModule}\\/\\b`))) {
      delete updatedCachedModules[cachedModuleKey];
      console.log(`Deleted ${cachedModuleKey} from cache`);
    }
  });
  return updatedCachedModules;
};
