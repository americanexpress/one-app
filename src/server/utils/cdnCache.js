/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'node:path';
import fs, { promises as fsPromises } from 'node:fs';
import chalk from 'chalk';

export const getUserHomeDirectory = () => process.env.HOME || process.env.USERPROFILE;
export const cacheFileName = '.one-app-module-cache';
export const oneAppDirectoryName = '.one-app';
export const oneAppDirectoryPath = path.join(getUserHomeDirectory(), oneAppDirectoryName);
export const oneAppModuleCachePath = path.join(oneAppDirectoryPath, cacheFileName);

// show cache size and how to delete info on start
export const showCacheInfo = async () => {
  try {
    const stats = await fsPromises.stat(oneAppModuleCachePath);
    const fileSizeInMB = stats.size / (1024 * 1024); // bytes to mb
    const separator = '*'.repeat(50);
    console.info(chalk.bold.cyanBright(separator));
    console.info(chalk.bold.cyanBright('CACHE INFORMATION'));
    console.info('File size of %s: %s', cacheFileName, chalk.bold.greenBright(fileSizeInMB.toFixed(2), 'MB'));
    console.info('To clear the cache, please delete this file:');
    console.info('    %s', chalk.bold.redBright(path.join('~', oneAppDirectoryName, cacheFileName)));
    console.info(chalk.bold.cyanBright(separator));
  } catch (error) {
    console.error('There was error checking file stat', error);
  }
};

// setup folder and file
export const setupCacheFile = async () => {
  try {
    await fsPromises.mkdir(oneAppDirectoryPath, { recursive: true });
    console.log(`Successfully created ${oneAppDirectoryPath}`);
    console.log(`Creating ${cacheFileName}`);
    try {
      await fsPromises.writeFile(oneAppModuleCachePath, JSON.stringify({}));
      console.log(`${cacheFileName} created successfully on ${oneAppModuleCachePath}`);
    } catch (error) {
      console.error(`Error creating ${cacheFileName} on ${oneAppModuleCachePath}, \n${error}`);
    }
  } catch (error) {
    console.error(`There was error creating ${oneAppDirectoryName} directory`);
  }
};

// gets cached module from ~/.one-app/.one-app-module-cache
export const getCachedModuleFiles = () => {
  if (!fs.existsSync(oneAppModuleCachePath)) {
    setupCacheFile();
    return {};
  }

  try {
    showCacheInfo();
    const cachedContent = fs.readFileSync(oneAppModuleCachePath, 'utf8');
    return JSON.parse(cachedContent);
  } catch (error) {
    console.error('Could not parse JSON content', error);
    return {};
  }
};

let timerId = null;

export const writeToCache = (content, delay = 500) => {
  // added debounce
  clearTimeout(timerId);
  timerId = setTimeout(() => {
    fs.writeFile(oneAppModuleCachePath, JSON.stringify(content, null, 2), (error) => {
      if (error) {
        console.log(`There was an error updating content \n ${error}`);
      }
    });
    timerId = null;
  }, delay);
};

const stripVersion = (url) => {
  const parts = url.split('/');
  parts.splice(-2, 1);
  return parts.join('/');
};

export const removeExistingEntryIfConflicting = (url, cachedModuleFiles) => {
  const updatedCachedModules = { ...cachedModuleFiles };
  const strippedUrl = stripVersion(url);

  const matchingModule = Object.keys(cachedModuleFiles)
    .find((cachedUrl) => stripVersion(cachedUrl) === strippedUrl);

  if (matchingModule && matchingModule !== url) {
    delete updatedCachedModules[matchingModule];
  }
  return updatedCachedModules;
};
