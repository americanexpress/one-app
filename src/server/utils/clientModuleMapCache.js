/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

let cache = {};

/* Filters bundle types so only the required client moduleBundleType is returned
in ../server/middleware/sendHtml.js and reducing the html payload size */
function filterBundles(moduleMap, moduleBundleType) {
  return {
    ...moduleMap,
    modules: Object.entries(moduleMap.modules).reduce((acc, [moduleName, moduleBundles]) => (
      {
        ...acc,
        [moduleName]: {
          baseUrl: moduleBundles[moduleBundleType].url.replace(/[^/]+\.js$/i, ''),
          [moduleBundleType]: moduleBundles[moduleBundleType],
        },
      }
    ), {}),
  };
}

export function setClientModuleMapCache(moduleMap) {
  cache = {
    browser: filterBundles(moduleMap, 'browser'),
    legacyBrowser: filterBundles(moduleMap, 'legacyBrowser'),
  };
}

export function getClientModuleMapCache() {
  return cache;
}
