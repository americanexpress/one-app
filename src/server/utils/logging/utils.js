/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { STATUS_CODES } from 'http';

import chalk from 'chalk';

const coloredLevels = {
  error: chalk.red('error'),
  warn: chalk.yellow('warn'),
  log: chalk.blue('log'),
  info: chalk.gray('info'),
};

function printStatusCode(obj) {
  const { request: { statusCode } } = obj;

  if (statusCode === null) {
    return chalk.black.bgRed(statusCode);
  }

  if (statusCode < 300) {
    // 100s, 200s
    return chalk.green(statusCode);
  }
  if (statusCode < 400) {
    // 300s
    return chalk.blue(statusCode);
  }
  if (statusCode < 500) {
    // 400s
    return chalk.yellow(statusCode);
  }
  // 500s
  return chalk.red(statusCode);
}

function printStatusMessage(obj) {
  const { request: { statusText, statusCode } } = obj;

  return statusText || STATUS_CODES[statusCode] || chalk.black.bgRed('Timed Out?');
}

function printDurationTime(obj) {
  const { request: { timings: { duration } } } = obj;
  if (duration < 100) {
    return chalk.green(duration);
  }
  if (duration < 1e3) {
    return chalk.yellow(duration);
  }
  if (duration < 6e3) {
    return chalk.red(duration);
  }
  return chalk.black.bgRed(duration);
}

export {
  coloredLevels,
  printStatusCode,
  printStatusMessage,
  printDurationTime,
};
