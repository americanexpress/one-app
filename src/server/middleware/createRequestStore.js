/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createHolocronStore } from 'holocron';
import { fromJS, Set as iSet } from 'immutable';

import { renderStaticErrorPage } from './sendHtml';
import createEnhancer from '../../universal/enhancers';
import safeRequest from '../utils/safeRequest';
import { getServerStateConfig, getClientStateConfig } from '../utils/stateConfig';
import createSsrFetch from '../utils/createSsrFetch';
import { getClientModuleMapCache } from '../utils/clientModuleMapCache';

export default function createRequestStore(
  { reducers },
  { useBodyForBuildingTheInitialState = false } = {}
) {
  return (req, res, next) => {
    try {
      const serverConfig = getServerStateConfig();
      const clientConfig = getClientStateConfig();

      const initialServerState = fromJS({
        holocron: { loaded: iSet() },
        config: serverConfig,
      });

      const fetchClient = createSsrFetch({
        req,
        res,
      })(fetch);

      const enhancer = createEnhancer();
      const localsForBuildInitialState = {
        req: safeRequest(req, { useBodyForBuildingTheInitialState }),
        config: clientConfig,
      };
      const store = createHolocronStore({
        reducer: reducers,
        initialState: initialServerState.merge(
          fromJS(reducers.buildInitialState(localsForBuildInitialState))
        ),
        enhancer,
        localsForBuildInitialState,
        extraThunkArguments: { fetchClient },
      });

      // TODO: namespace?
      // use the store as a global for the request
      req.store = store; // eslint-disable-line no-param-reassign
      req.clientModuleMapCache = getClientModuleMapCache();
      return next();
    } catch (err) {
      console.error('error creating store for request', err);
      return renderStaticErrorPage(res);
    }
  };
}
