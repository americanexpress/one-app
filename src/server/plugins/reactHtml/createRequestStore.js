/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { createHolocronStore } from 'holocron';
import { fromJS, Set as iSet } from 'immutable';

import renderStaticErrorPage from './staticErrorPage';
import createEnhancer from '../../../universal/enhancers';
import safeRequest from '../../utils/safeRequest';
import { getServerStateConfig, getClientStateConfig } from '../../utils/stateConfig';
import createSsrFetch from '../../utils/createSsrFetch';
import { getClientModuleMapCache } from '../../utils/clientModuleMapCache';

const createRequestStore = (
  request, reply,
  { reducers }
) => {
  // fastify.decorateRequest('store', null);
  // fastify.decorateRequest('clientModuleMapCache', null);

  try {
    const serverConfig = getServerStateConfig();
    const clientConfig = getClientStateConfig();

    const initialServerState = fromJS({
      holocron: { loaded: iSet() },
      config: serverConfig,
    });

    const fetchClient = createSsrFetch({
      req: request.raw,
      res: reply.raw,
    })(fetch);

    console.log('--request.method', request.method);
    console.log('--request.body', request.body);
    console.log('--request.raw.body', request.raw.body);

    const enhancer = createEnhancer();
    const localsForBuildInitialState = {
      req: safeRequest(request, { useBodyForBuildingTheInitialState: request.method.toLowerCase() === 'post' }),
      config: clientConfig,
    };

    const store = createHolocronStore({
      reducer: reducers,
      initialState: initialServerState.merge(
        fromJS(reducers.buildInitialState(localsForBuildInitialState))
      ),
      enhancer,
      localsForBuildInitialState,
      extraThunkArguments: { fetchClient },
    });

    // TODO: namespace?
    // use the store as a global for the request
    request.store = store; // eslint-disable-line no-param-reassign
    request.clientModuleMapCache = getClientModuleMapCache();
  } catch (err) {
    console.error('error creating store for request', err);
    // TODO: migrate `renderStaticErrorPage`
    return renderStaticErrorPage(reply);
  }
};

export default createRequestStore;
