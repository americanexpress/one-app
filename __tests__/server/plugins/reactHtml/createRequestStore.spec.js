/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import combineReducers from '@americanexpress/vitruvius/immutable';
import holocron from 'holocron';
import { fromJS } from 'immutable';
import renderStaticErrorPage from '../../../../src/server/plugins/reactHtml/staticErrorPage';
import createRequestStore from '../../../../src/server/plugins/reactHtml/createRequestStore';
import { setClientModuleMapCache } from '../../../../src/server/utils/clientModuleMapCache';

jest.mock('../../../../src/server/plugins/reactHtml/staticErrorPage');
jest.mock('holocron', () => {
  const actualHolocron = jest.requireActual('holocron');
  return {
    ...actualHolocron,
    createHolocronStore: jest.fn(actualHolocron.createHolocronStore),
    getModuleMap: jest.fn(),
  };
});

holocron.getModuleMap.mockImplementation(() => fromJS({
  modules: {
    'test-root': {
      baseUrl: 'https://example.com/cdn/test-root/2.2.2/',
      node: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.node.js',
        integrity: '4y45hr',
      },
      browser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.browser.js',
        integrity: 'nggdfhr34',
      },
      legacyBrowser: {
        url: 'https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js',
        integrity: '7567ee',
      },
    },
  },
}));

setClientModuleMapCache(holocron.getModuleMap().toJS());

describe('createRequestStore', () => {
  let request;
  let reply;
  let reducers;
  const span = { end: jest.fn() };
  const tracer = { startSpan: jest.fn(() => span) };

  beforeAll(() => {
    global.fetch = () => Promise.resolve({ data: 'data' });
    holocron.createHolocronStore.mockClear();
  });

  afterAll(() => {
    global.fetch = undefined;
  });

  beforeEach(() => {
    jest.clearAllMocks();

    request = {
      openTelemetry: () => ({ tracer }),
      decorateRequest: jest.fn(),
      headers: {
        'correlation-id': 'abc123',
      },
      url: '/',
      raw: {},
      method: 'get',
      log: { error: jest.fn() },
    };

    reply = {
      raw: {},
    };

    reducers = jest.fn(
      (...args) => combineReducers({
        appReducer: jest.fn((v) => v || true),
        config: jest.fn((v) => v || true),
      })(...args)
    );
    reducers.buildInitialState = jest.fn(() => fromJS({ appReducer: 'fizzy' }));
  });

  it('should add a store to the request object', () => {
    createRequestStore(request, reply, { reducers });

    expect(request.log.error).not.toHaveBeenCalled();
    expect(request.store).toBeTruthy();
  });

  it('should add the client holocron module map cache to the request object', () => {
    createRequestStore(request, reply, { reducers });

    expect(request.clientModuleMapCache).toMatchSnapshot();
  });

  it('should send the static error page when there is an error', () => {
    createRequestStore(request, reply, { reducers: null });

    expect(request.log.error).toHaveBeenCalled();
    expect(renderStaticErrorPage).toHaveBeenCalledWith(request, reply);
  });

  describe('fetch', () => {
    it('should set the store up with fetchClient', async () => {
      createRequestStore(request, reply, { reducers });
      const {
        extraThunkArguments: { fetchClient },
      } = holocron.createHolocronStore.mock.calls[0][0];

      expect(fetchClient).toBeInstanceOf(Function);
    });
  });

  describe('useBodyForBuildingTheInitialState', () => {
    it('uses the request body as the locals for initial state when useBodyForBuildingTheInitialState is true', () => {
      request.method = 'post';
      request.body = { some: 'form data', that: 'was posted' };

      createRequestStore(request, reply, { reducers });

      expect(reducers.buildInitialState).toHaveBeenCalledTimes(1);
      expect(reducers.buildInitialState.mock.calls[0][0]).toHaveProperty('req.body.some', 'form data');
      expect(reducers.buildInitialState.mock.calls[0][0]).toHaveProperty('req.body.that', 'was posted');
    });

    it('does not use the request body as the locals for initial state when useBodyForBuildingTheInitialState is not given', () => {
      request.body = { some: 'other form data', that: 'was acquired' };
      createRequestStore(request, reply, { reducers });

      expect(reducers.buildInitialState).toHaveBeenCalledTimes(1);
      expect(reducers.buildInitialState.mock.calls[0][0]).not.toHaveProperty('req.body');
    });
  });

  it('should pass enhancedFetch into createHolocronStore', () => {
    createRequestStore(request, reply, { reducers });

    expect(holocron.createHolocronStore.mock.calls[0][0]).toHaveProperty('extraThunkArguments.fetchClient');
  });

  describe('tracer', () => {
    it('should create a span', () => {
      createRequestStore(request, reply, { reducers });

      expect(tracer.startSpan).toHaveBeenCalledWith('createRequestStore');
      expect(span.end).toHaveBeenCalled();
    });

    it('should still end the span if an error is thrown', () => {
      createRequestStore(request, reply, { reducers: null });

      expect(tracer.startSpan).toHaveBeenCalledWith('createRequestStore');
      expect(request.log.error).toHaveBeenCalled();
      expect(renderStaticErrorPage).toHaveBeenCalledWith(request, reply);
      expect(span.end).toHaveBeenCalled();
    });
  });
});
