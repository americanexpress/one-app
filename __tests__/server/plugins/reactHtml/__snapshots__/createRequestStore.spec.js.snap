// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createRequestStore should add the client holocron module map cache to the request object 1`] = `
{
  "browser": {
    "modules": {
      "test-root": {
        "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
        "browser": {
          "integrity": "nggdfhr34",
          "url": "https://example.com/cdn/test-root/2.2.2/test-root.browser.js",
        },
      },
    },
  },
  "legacyBrowser": {
    "modules": {
      "test-root": {
        "baseUrl": "https://example.com/cdn/test-root/2.2.2/",
        "legacyBrowser": {
          "integrity": "7567ee",
          "url": "https://example.com/cdn/test-root/2.2.2/test-root.legacy.browser.js",
        },
      },
    },
  },
}
`;
